driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -343,13 +338,25 @@ private void killCursor() {
     private void killCursor(final Connection connection) {
         if (serverCursor != null) {
             notNull("connection", connection);
-            if (serverIsAtLeastVersionThreeDotTwo(connection.getDescription())) {
-                connection.command(namespace.getDatabaseName(), asKillCursorsCommandDocument(), NO_OP_FIELD_NAME_VALIDATOR,
-                        ReadPreference.primary(), new BsonDocumentCodec(), connectionSource.getSessionContext());
-            } else {
-                connection.killCursor(namespace, singletonList(serverCursor.getId()));
+            try {
+                if (serverIsAtLeastVersionThreeDotTwo(connection.getDescription())) {
+                    connection.command(namespace.getDatabaseName(), asKillCursorsCommandDocument(), NO_OP_FIELD_NAME_VALIDATOR,
+                            ReadPreference.primary(), new BsonDocumentCodec(), connectionSource.getSessionContext());
+                } else {
+                    connection.killCursor(namespace, singletonList(serverCursor.getId()));
+                }
+            } catch (MongoException e) {
+                // Ignore exceptions from calling killCursor
+            } finally {
+                serverCursor = null;
             }
-            serverCursor = null;
+        }
+    }
+
+    private void checkServerCursorAndReleaseConnectionSource() {
```

Comment:
Nit: consider renaming to `releaseConnectionSourceIfNoServerCursor`

Reply:
Done